src/System.Memory/src/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs

```diff
@@ -21,13 +22,14 @@ private sealed class ArrayMemoryPoolBuffer : OwnedMemory<T>
             public ArrayMemoryPoolBuffer(int size)
             {
                 _array = ArrayPool<T>.Shared.Rent(size);
+                _refCount = 1;
             }
 
             public sealed override int Length => _array.Length;
 
             public sealed override bool IsDisposed => _array == null;
 
-            protected sealed override bool IsRetained => _refCount > 0;
+            protected sealed override bool IsRetained => Volatile.Read(ref _refCount) > 0;
```

Comment:
@stephentoub, does it need to be interlocked read?

Reply:
> does it need to be interlocked read?

No, it's only 32-bit, so there's no risk of tearing and Interlocked.Read isn't needed.  Volatile.Read may not even be necessary depending on what it's used for, but it's unlikely to hurt and so might as well be used.

Reply:
Ah, I missed that it's an int.